text/2020-03-07-copr-chunk.md

```diff
@@ -0,0 +1,214 @@
+    // TODO
+}
+
+pub trait FixedSizeChunkedVec<T>: ChunkedVec<T> { /* TODO */ }
+
+pub trait VarSizeChunkedVec<T>: ChunkedVec<T> { /* TODO */ }
+```
+
+
+### Rpn Expression
+
+Currently, An expression receives `&[RpnStackNode<'_>],` as input args, the definition of `RpnStackNode` is:
+
+```rust
+#[derive(Debug)]
+pub enum RpnStackNode<'a> {
+    /// Represents a scalar value. Comes from a constant node in expression list.
+    Scalar {
+        value: &'a ScalarValue,
+        field_type: &'a FieldType,
+    },
+
+    /// Represents a vector value. Comes from a column reference or evaluated result.
+    Vector {
+        value: RpnStackNodeVectorValue<'a>,
+        field_type: &'a FieldType,
+    },
+}
+```
+
+For non constant value, a column of data is represented by `RpnStackNodeVectorValue`:
+
+```rust
+#[derive(Debug)]
+pub enum RpnStackNodeVectorValue<'a> {
+    Generated {
+        // TODO: Maybe box it can be faster.
+        physical_value: VectorValue,
+    },
+    Ref {
+        physical_value: &'a VectorValue,
+        logical_rows: &'a [usize],
+    },
+}
+```
+
+Currently, the definition of `VectorValue` is:
+
+```rust
+#[derive(Debug, PartialEq, Clone)]
+pub enum VectorValue {
+    Int(Vec<Option<Int>>),
+    Real(Vec<Option<Real>>),
+    Decimal(Vec<Option<Decimal>>),
+    // TODO: We need to improve its performance, i.e. store strings in adjacent memory places
+    Bytes(Vec<Option<Bytes>>),
+    DateTime(Vec<Option<DateTime>>),
+    Duration(Vec<Option<Duration>>),
+    Json(Vec<Option<Json>>),
+}
+```
+
+We should modify the `VectorValue` to:
+
+```rust
+#[derive(Debug, PartialEq, Clone)]
+pub enum VectorValue {
+    Int(ChunkedVec<Int>),
+    Real(ChunkedVec<Real>),
+    Decimal(ChunkedVec<Decimal>),
+    Bytes(ChunkedVec<Bytes>),
+    DateTime(ChunkedVec<DateTime>),
+    Duration(ChunkedVec<Duration>),
+    Json(ChunkedVec<Json>),
+}
+```
+
+`ChunkedVec` implemented [chunk column format](https://pingcap.com/blog-cn/tidb-source-code-reading-10/).
+
+The original definition of `Bytes` is `type Bytes = Vec<u8>`, we have to introduce `type BytesRef<'a> = &'a [u8]` as type of input arguments due to the memory of `Bytes` is actually owned by `ChunkedVec`.
+
+The `ScalarRef` is impossible to keep compatibility.
+
+```rust
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum ScalarValueRef<'a> {
+    // Fixed sized.
+    Int(&'a Option<Int>),
+    // ...
+    Bytes(Option<&Bytes>),
+    Json(Option<&Json>),
+}
+```
+
+We have to modify it to:
+
+```rust
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum ScalarValueRef<'a> {
+    // Fixed sized.
+    Int(Option<&'a Int>),
+    // ...
+    Bytes(Option<BytesRef>),
+    Json(Option<JsonRef>),
+}
+```
+
+We also have to modify all rpn_function signatures.
+
+For fixed size type:
+
+```rust
+// From
+#[rpn_fn]
+pub fn logical_and(lhs: &Option<i64>, rhs: &Option<i64>) -> Result<Option<i64>>;
+
+// To
+#[rpn_fn]
+pub fn logical_and(lhs: &i64, rhs: &i64) -> Result<Option<i64>>;
```

Comment:
Here is a question. For fixed sized type, is it better to use value type instead of ref? almost all fixed type <= 8 byte.